File: Objective-C/Search/Controller/ViewController.m
//
//  ViewController.m
//  SearchBothiOS7AndiOS8
//
//  Created by John Lima on 5/24/15.
//  Copyright (c) 2015 limadeveloper. All rights reserved.
//

#import "ViewController.h"
#import "Model.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate, UISearchBarDelegate, UISearchControllerDelegate, UISearchResultsUpdating>

@end

@implementation ViewController

#pragma mark - Properties
BOOL searchEnable = NO;
UISearchController *searchController;
NSArray *tableData;
NSMutableArray *searchData;
NSIndexPath *selectedIndexPath;
UIRefreshControl *refreshControll;

#pragma mark - View LifeCycle
-(void)viewDidLoad {
    [super viewDidLoad];
    
    [self createSearch];
    [self setupRefresh];
    
    [self getData];
    [self updateUI];
}

-(void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    if (searchEnable) {
        [self searchBarCancelButtonClicked:searchController.searchBar];
        [self clearSearch];
    }
}

#pragma mark - Actions
-(void)updateUI {
    
    UIView *footer = [[UIView alloc] initWithFrame:CGRectZero];
    
    self.tableView.tableFooterView = footer;
    self.tableView.backgroundColor = [UIColor whiteColor];
    self.tableView.allowsSelection = tableData.count > 0;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
    [self.tableView reloadData];
}

-(void)getData {
    
    Model *model = [[Model alloc] init];
    
    tableData = [[NSArray alloc] init];
    tableData = [model getData];
    
    [refreshControll endRefreshing];
    [self.tableView reloadData];
}

-(void)createSearch {
    
    searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
    
    [searchController loadViewIfNeeded];
    
    searchController.searchResultsUpdater = self;
    
    [searchController.searchBar sizeToFit];
    [searchController.searchBar setBarTintColor:[UIColor purpleColor]];
    [searchController.searchBar setTintColor:[UIColor whiteColor]];
    
    [self.viewSearch addSubview:searchController.searchBar];
    
    searchController.delegate = self;
    searchController.dimsBackgroundDuringPresentation = false;
    searchController.hidesNavigationBarDuringPresentation = false;
    searchController.searchBar.delegate = self;
}

-(void)setupRefresh {
    
    refreshControll = [[UIRefreshControl alloc] init];
    refreshControll.tintColor = [UIColor purpleColor];
    
    [refreshControll addTarget:self action:@selector(refreshData) forControlEvents:UIControlEventValueChanged];
    
    [self.tableView addSubview:refreshControll];
}

-(void)clearSearch {
    
    searchData = [[NSMutableArray alloc] init];
    searchEnable = NO;
    
    [self.tableView reloadData];
}

-(void)refreshData {
    [self getData];
    [self updateUI];
    [self clearSearch];
}

#pragma mark - Search Delegates

-(void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    
    NSString *searchText = searchController.searchBar.text;
    
    if (searchText == nil) {
        searchData = [[NSMutableArray alloc] init];
    }else {
    
        searchData = [[NSMutableArray alloc] init];
        
        for (Model *model in tableData) {
            if ([model.name containsString:searchText]) {
                [searchData addObject:model];
            }
        }
    }
    
    searchEnable = searchController.searchBar.showsCancelButton;
    
    [self.tableView reloadData];
}

-(void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    
    searchBar.text = nil;
    searchBar.showsCancelButton = NO;
    
    [searchBar resignFirstResponder];
    
    self.tableView.allowsSelection = YES;
}

-(void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    [self searchBarCancelButtonClicked:searchBar];
}

-(void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar {
    [self clearSearch];
    self.tableView.allowsSelection = NO;
    searchBar.showsCancelButton = YES;
}

#pragma mark - TableView DataSource
-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSArray *data = searchEnable ? searchData : tableData;
    return data.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    
    NSArray *data = searchEnable ? searchData : tableData;
    
    if (data.count > 0) {
        cell.textLabel.text = [(Model *)data[indexPath.row] name];
        cell.textLabel.textColor = [UIColor purpleColor];
    }
    
    return cell;
}

#pragma mark - TableView Delegate
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    [searchController setActive:NO];
    
    selectedIndexPath = indexPath;
}

@end

File: Swift/Search/Controller/ViewController.swift
//
//  ViewController.swift
//  Search
//
//  Created by John Lima on 19/04/17.
//  Copyright © 2017 John Lima. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    // MARK: - Properties
    @IBOutlet fileprivate weak var tableView: UITableView!
    @IBOutlet fileprivate weak var viewSearch: UIView!
    
    let model = Model()
    let cellName = "cell"
    var searchEnable = false
    var searchController: UISearchController?
    var tableData: Array<Model>?
    var searchData: Array<Model>?
    var selectedIndexPath: IndexPath?
    var refreshControll: UIRefreshControl?
    
    // MARK: - View LifeCycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        createSearch()
        setupRefresh()
        getData()
        updateUI()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        guard let searchBar = searchController?.searchBar, searchEnable else { return }
        
        searchBarCancelButtonClicked(searchBar)
        clearSearch()
    }
    
    // MARK: - Actions
    fileprivate func createSearch() {
    
        searchController = UISearchController(searchResultsController: nil)
        searchController?.loadViewIfNeeded()
        searchController?.searchResultsUpdater = self
        
        searchController?.searchBar.sizeToFit()
        searchController?.searchBar.barTintColor = .purple
        searchController?.searchBar.tintColor = .white
        
        if let searchBar = searchController?.searchBar {
            viewSearch.addSubview(searchBar)
        }
        
        searchController?.delegate = self
        
        searchController?.dimsBackgroundDuringPresentation = false
        searchController?.hidesNavigationBarDuringPresentation = false
        searchController?.searchBar.delegate = self
    }
    
    fileprivate func setupRefresh() {
        
        refreshControll = UIRefreshControl()
        refreshControll?.tintColor = .purple
        refreshControll?.addTarget(self, action: #selector(refreshData), for: .valueChanged)
        
        guard let refreshControll = refreshControll else { return }
        
        tableView.addSubview(refreshControll)
    }
    
    fileprivate func getData() {
    
        tableData = model.getData()
        
        refreshControll?.endRefreshing()
        tableView.reloadData()
    }
    
    fileprivate func updateUI() {
        
        let footer = UIView(frame: .zero)
        
        tableView.tableFooterView = footer
        tableView.backgroundColor = .white
        tableView.allowsSelection = (tableData ?? []).count > 0
        tableView.delegate = self
        tableView.dataSource = self
        tableView.reloadData()
    }
    
    fileprivate func clearSearch() {
        searchData = []
        searchEnable = false
        tableView.reloadData()
    }
    
    dynamic fileprivate func refreshData() {
        getData()
        updateUI()
        clearSearch()
    }
}

extension ViewController: UISearchBarDelegate, UISearchControllerDelegate, UISearchResultsUpdating {
    
    func updateSearchResults(for searchController: UISearchController) {
        
        searchData = []
        
        guard let searchText = searchController.searchBar.text, !searchText.isEmpty else { return }
        
        for model in (tableData ?? []) {
            if (model.name ?? "").contains(searchText) {
                searchData?.append(model)
            }
        }
        
        searchEnable = searchController.searchBar.showsCancelButton
        tableView.reloadData()
    }
    
    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        
        searchBar.text = nil
        searchBar.showsCancelButton = false
        searchBar.resignFirstResponder()
        
        tableView.allowsSelection = true
    }
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBarCancelButtonClicked(searchBar)
    }
    
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        clearSearch()
        tableView.allowsSelection = false
        searchBar.showsCancelButton = true
    }
}

extension ViewController: UITableViewDataSource, UITableViewDelegate {

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let data = searchEnable ? searchData : tableData
        return (data ?? []).count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: cellName, for: indexPath)
        
        guard let data = searchEnable ? searchData : tableData else { return cell }
        
        cell.textLabel?.text = data[indexPath.row].name
        cell.textLabel?.textColor = .purple
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        tableView.deselectRow(at: indexPath, animated: true)
        
        searchController?.isActive = false
        selectedIndexPath = indexPath
    }
}

